File: 739-daily-temperatures/solution.ts
function dailyTemperatures(temperatures: number[]): number[] {
  const stack: [number, number][] = [];
  const answer = temperatures.map(() => 0);
  for (let i = 0; i < temperatures.length; i++) {
    while (temperatures[i] > stack[stack.length - 1]?.[1]) {
      const [idx] = stack.pop() as [number, number];
      answer[idx] = i - idx;
    }
    stack.push([i, temperatures[i]]);
  }
  return answer;
}

File: 739-daily-temperatures/solution.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        vector<int> ans(n, 0);
        stack<pair<int, int>> sk;
        for(int i = 0; i < n; i++) {
            int temp = temperatures[i];
            while(!sk.empty() && sk.top().first < temp) {
                auto [t, idx] = sk.top();
                sk.pop();
                ans[idx] = i - idx;
            }
            sk.push({temp, i});
        }
        return ans;
    }
};
